### PR TITLE
s3: resolve bucket ARN in replication config references

### LINKED ISSUE #745: S3 Bucket replication rules for bucket ref resolves to external name instead of bucket ARN
### What happened?
When configuring S3 Bucket replication rules and using the bucketRef  to select a bucket by name, the external name is supplied instead of the bucket ARN in the resolved resource.

The error produced:
```
  Warning  CannotUpdateExternalResource  14s (x5 over 23s)  managed/bucket.s3.aws.crossplane.io  cannot create or update: cannot put Bucket replication: InvalidArgument: Invalid bucket ARN
```

It looks like the wrong extraction function is supplied to the bucket references: https://github.com/crossplane/provider-aws/blob/master/apis/s3/v1beta1/referencers.go#L106

### What environment did it happen in?
Crossplane version: 1.2.3
aws-provider version 0.19.0

### PR DESCRIPTION
**Pull request by chlunde**
### Description of your changes

The replication destination bucket should resolve to an ARN, not a bucket name.

Fixes #745

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
With blood, sweat and tears. Fixed three other issues first.

This is a "minimal" configuration (excluding role and policy):
```yaml
---
apiVersion: s3.aws.crossplane.io/v1beta1
kind: Bucket
metadata:
  annotations:
    crossplane.io/external-name: repl-src.2021-06-07
  name: repl-src
spec:
  deletionPolicy: Delete
  forProvider:
    acl: private
    locationConstraint: eu-north-1
    paymentConfiguration:
      payer: BucketOwner
    serverSideEncryptionConfiguration:
      rules:
      - applyServerSideEncryptionByDefault:
          sseAlgorithm: AES256
    versioningConfiguration:
      status: Enabled
    replicationConfiguration:
      role: arn:aws:iam::...:role/s3-replication-test # assumable by s3 svc, I just tested with an s3 admin role
      rules:
      - destination:
          storageClass: STANDARD
          bucketRef:
            name: repl-dest
        deleteMarkerReplication:
          status: Disabled
        priority: 0
        id: rule-1
        status: Enabled
---
apiVersion: s3.aws.crossplane.io/v1beta1
kind: Bucket
metadata:
  annotations:
    crossplane.io/external-name: repl-dest.2021-06-07
  name: repl-dest
spec:
  deletionPolicy: Delete
  forProvider:
    acl: private
    locationConstraint: eu-north-1
    paymentConfiguration:
      payer: BucketOwner
    serverSideEncryptionConfiguration:
      rules:
      - applyServerSideEncryptionByDefault:
          sseAlgorithm: AES256
    versioningConfiguration:
      status: Enabled

```
Question: Do you prefer the function returning a function for BucketARN?

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by AaronME — CHANGES_REQUESTED** [review]
@chlunde, thank you for addressing https://github.com/crossplane/provider-aws/issues/745

I'm getting the following error in my tests:

```bash
"cannot create or update: cannot put Bucket replication: MalformedXML: The XML you provided was not well-formed or did not validate against our published schema"
```

I used the following configuration:
```yaml
---
apiVersion: s3.aws.crossplane.io/v1beta1
kind: Bucket
metadata:
  name: repl-src
spec:
  deletionPolicy: Delete
  forProvider:
    acl: private
    locationConstraint: us-west-2
    paymentConfiguration:
      payer: BucketOwner
    serverSideEncryptionConfiguration:
      rules:
      - applyServerSideEncryptionByDefault:
          sseAlgorithm: AES256
    versioningConfiguration:
      status: Enabled
    replicationConfiguration:
      role: arn:aws:iam::...:role/role-name # assumable by s3 svc, I just tested with an s3 admin role
      rules:
      - destination:
          storageClass: STANDARD_IA
          bucketRef:
            name: repl-dest
        deleteMarkerReplication:
          Status: Disabled
        priority: 0
        id: rule-1
        status: Enabled
  providerConfigRef:
    name: example
---
apiVersion: s3.aws.crossplane.io/v1beta1
kind: Bucket
metadata:
  name: repl-dest
spec:
  deletionPolicy: Delete
  forProvider:
    acl: private
    locationConstraint: us-west-2
    paymentConfiguration:
      payer: BucketOwner
    serverSideEncryptionConfiguration:
      rules:
      - applyServerSideEncryptionByDefault:
          sseAlgorithm: AES256
    versioningConfiguration:
      status: Enabled
  providerConfigRef:
    name: example

```

**Comment by chlunde** [comment]
Thanks for testing and reviewing this and the other PRs! I think you need a filter, or the filter PR, but I'm not certain.

I can check on monday.

**Review by AaronME — CHANGES_REQUESTED** [review]
@chlunde, after merging the filter fix I successfully created the replicating bucket pair. Nice work!

Before merging, could you please:
1. Include the typo fix from https://github.com/crossplane/provider-aws/pull/766 in this PR.
1. Update examples/s3/ with your test cases so we can run conformance testing on these changes.
1. Rebase on latest master.

Thanks again for this excellent contribution!

**Comment by chlunde** [comment]
```
cannot create or update: cannot put Bucket replication: InvalidRequest: Versioning must be 'Enabled' on the bucket to apply a replication configuration
```

*sigh* - apparently we need to re-order those calls

Also, I'm back to malformed policy unless i specify a filter, I don't understand why, because the SDK version is the same as before. Added filter to the example.

File: examples/s3/bucket.yaml [hunk 1]
```diff
@@ -46,5 +46,44 @@ spec:
...
+        name: somerole
+      rules:
+        - destination:
+            storageClass: STANDARD
+            bucketRef:
+              name: repl-dest
+          deleteMarkerReplication:
+            status: Disabled
+          filter:
+            prefix: ""
```
Comment (AaronME) [inline examples/s3/bucket.yaml:60]: @chlunde This works for me both with and without this filter.

**Review by AaronME — APPROVED** [review]
@chlunde This works for me!

Thanks!

File: examples/s3/bucket.yaml (hunk 1 above)
Comment (chlunde) [reply examples/s3/bucket.yaml:60]: cool, do you want me to remove it or keep it as it is?

**Pull request merged**